llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

```diff
@@ -1172,6 +1209,21 @@ bool SIPeepholeSDWA::convertToSDWA(MachineInstr &MI,
     SDWAInst->tieOperands(PreserveDstIdx, SDWAInst->getNumOperands() - 1);
   }
 
+  return SDWAInst.getInstr();
+}
+
+bool SIPeepholeSDWA::convertToSDWA(MachineInstr &MI,
+                                   const SDWAOperandsVector &SDWAOperands) {
+  LLVM_DEBUG(dbgs() << "Convert instruction:" << MI);
+
+  MachineInstr *SDWAInst;
+  if (TII->isSDWA(MI.getOpcode())) {
+    SDWAInst = MI.getParent()->getParent()->CloneMachineInstr(&MI);
```

Comment:
Do we really need the clone? Maybe we can handle code removal just for the case with no pre-existing sdwa

Reply:
I don't think we can avoid the clone. I actually gave it a try first - using `SDWAInst = &MI` if MI is already an SDWAInst and making the code removal conditioned on `SDWAInst != &MI`. This made, for instance, the "MC/AMDGPU/gfx10_asm_vopc_sdwa.s." test hang. It makes sense that we need to operate on a copy if you look at the `convertToSDWA` implementations for the different Operand types. They may change the MI. If the conversion fails for one of them, we need to able to revoke those changes. The base version of this code also always operates on a copy which is created in the code that has been moved into `createSDWAVersion` by this PR.